### PR TITLE
fix: allow resizing to min required size for performance plus disks

### LINKED ISSUE #1949: misleading message for performancePlus disk size
**What happened**:

Creating disk with 512Gi size result in error saying that the disk needs to be at least 512GB for performancePlus.

**What you expected to happen**:

1) receive correct error message saying it's at least 513Gi, which is actually documented correctly here: https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-performance?tabs=azure-cli#enable-performance-plus

2) driver should resize the volume accordingly if the performancePlus feature is enabled, similar to what Azure File driver does with premium fileshares: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/794f5fa89560cfb4b3a183bccdd73b8f3c4bd8f0/pkg/azurefile/controllerserver.go#L358-L363

**How to reproduce it**:

1. Create SC with `enablePerformancePlus` parameter set to `true`:

```
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-perfplus
parameters:
  skuname: Premium_LRS
  enablePerformancePlus: "true"
provisioner: disk.csi.azure.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

2. Create PVC with exactly 512Gi

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mypvc-1
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 512Gi
  storageClassName: "csi-perfplus"
```

3. Observe `ProvisioningFailed` message:  

` Warning  ProvisioningFailed    <invalid> (x5 over <invalid>)  disk.csi.azure.com_wduan0810manual-b5dng-master-1_d7a29bbf-3f49-4207-af33-056e0814f6e2  failed to provision volume with StorageClass "managed-csi-test-28-sssdlrs-enableperformanceplus": rpc error: code = Internal desc = Retriable: false, RetryAfter: 0s, HTTPStatusCode: 400, RawError: {
  "error": {
    "code": "BadRequest",
    "message": "The performancePlus flag can only be set on disks at least 512 GB in size."
  }
}
`

4. Retry with 513Gi PVC

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mypvc-2
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 513Gi
  storageClassName: "csi-perfplus"
```

6. Check that disk created successfully and performancePlus is set

```
$ az disk show --resource-group rbednar-mycluster-01-4c6xn-rg --name `oc get pv/pvc-5ca12393-504f-40d2-b9ca-1bcd7198ee18 -o custom-columns=:.metadata.name --no-headers` | jq '.creationData.performancePlus'
true
```

**Anything else we need to know?**:

**Environment**:

- CSI Driver version: 1.28
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by RomanBednar**
**What type of PR is this?**
/kind bug
/kind feature

**What this PR does / why we need it**:

See #1949 for detailed description. This patch would reduce the number of errors users might get when creating performancePlus disks.

**Which issue(s) this PR fixes**:

Fixes #1949

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

Patch verified with:

1) Create StorageClass:
```
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-perfplus
parameters:
  skuname: Premium_LRS
  enablePerformancePlus: "true"
provisioner: disk.csi.azure.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

2) Create PVC:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mypvc-enableperformanceplus
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  storageClassName: "csi-perfplus"
```

3) Warning log message:

`W0829 10:46:37.457074       1 controllerserver.go:103] using PerformancePlus, increasing requested disk size from 5GiB to 513GiB (minimal size for PerformancePlus feature)`

4) Check disk was created with performancePlus enabled:
```
$ az disk show --resource-group rbednar-mycluster-01-4c6xn-rg --name `oc get pv/pvc-3795b4cd-c105-462d-9a95-99e6ce35282b -o custom-columns=:.metadata.name --no-headers` | jq '.creationData.performancePlus'
true
```

**Release note**:
```
When `enablePerformancePlus` parameter is set to `"true"` volumes will be automatically resized to 513Gi if requested size was smaller.
```

### CODE REVIEW

File: pkg/azuredisk/controllerserver.go [hunk 1]
```diff
@@ -98,7 +98,13 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
+
+	if diskParams.PerformancePlus != nil && *diskParams.PerformancePlus {
```
Comment (andyzhangx) [inline pkg/azuredisk/controllerserver.go:102]: `if ... && requestGiB <  consts.PerformancePlusMinimumDiskSizeGiB`

Comment (RomanBednar) [reply pkg/azuredisk/controllerserver.go:102]: Ah, sure, fixed.

Comment (andyzhangx) [reply pkg/azuredisk/controllerserver.go:102]: still not fixed?

Comment (RomanBednar) [reply pkg/azuredisk/controllerserver.go]: sorry, now it should be ok

File: pkg/azuredisk/controllerserver.go [hunk 2]
```diff
@@ -98,7 +98,13 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
+
+	if diskParams.PerformancePlus != nil && *diskParams.PerformancePlus && requestGiB < consts.PerformancePlusMinimumDiskSizeGiB {
+		klog.Warningf("using PerformancePlus, increasing requested disk size from %vGiB to %vGiB (minimal size for PerformancePlus feature)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
+		requestGiB = consts.PerformancePlusMinimumDiskSizeGiB
+	}
 	if requestGiB < consts.MinimumDiskSizeGiB {
+		klog.Infof("increasing requested disk size from %vGiB to %vGiB (minimal disk size)")
```
Comment (andyzhangx) [inline pkg/azuredisk/controllerserver.go:107]: append `requestGiB, consts.MinimumDiskSizeGiB`

File: pkg/azuredisk/controllerserver_v2.go [hunk 3]
```diff
@@ -89,7 +89,13 @@ func (d *DriverV2) CreateVolume(ctx context.Context, req *csi.CreateVolumeReques
 	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
+
+	if diskParams.PerformancePlus != nil && *diskParams.PerformancePlus && requestGiB < consts.PerformancePlusMinimumDiskSizeGiB {
+		klog.Warningf("using PerformancePlus, increasing requested disk size from %vGiB to %vGiB (minimal size for PerformancePlus feature)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
+		requestGiB = consts.PerformancePlusMinimumDiskSizeGiB
+	}
 	if requestGiB < consts.MinimumDiskSizeGiB {
+		klog.Infof("increasing requested disk size from %vGiB to %vGiB (minimal disk size)")
```
Comment (andyzhangx) [inline pkg/azuredisk/controllerserver_v2.go:98]: append `requestGiB, consts.MinimumDiskSizeGiB`

Comment (RomanBednar) [reply pkg/azuredisk/controllerserver_v2.go]: Done.

File: pkg/azuredisk/controllerserver.go [hunk 4]
```diff
@@ -98,7 +98,13 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
+
+	if diskParams.PerformancePlus != nil && *diskParams.PerformancePlus && requestGiB < consts.PerformancePlusMinimumDiskSizeGiB {
+		klog.Warningf("using PerformancePlus, increasing requested disk size from %vGiB to %vGiB (minimal size for PerformancePlus feature)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
+		requestGiB = consts.PerformancePlusMinimumDiskSizeGiB
+	}
 	if requestGiB < consts.MinimumDiskSizeGiB {
+		klog.Infof("increasing requested disk size from %vGiB to %vGiB (minimal disk size)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
```
Comment (andyzhangx) [inline pkg/azuredisk/controllerserver.go:107]: it should be `MinimumDiskSizeGiB`

File: pkg/azuredisk/controllerserver_v2.go [hunk 5]
```diff
@@ -89,7 +89,13 @@ func (d *DriverV2) CreateVolume(ctx context.Context, req *csi.CreateVolumeReques
 	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
+
+	if diskParams.PerformancePlus != nil && *diskParams.PerformancePlus && requestGiB < consts.PerformancePlusMinimumDiskSizeGiB {
+		klog.Warningf("using PerformancePlus, increasing requested disk size from %vGiB to %vGiB (minimal size for PerformancePlus feature)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
+		requestGiB = consts.PerformancePlusMinimumDiskSizeGiB
+	}
 	if requestGiB < consts.MinimumDiskSizeGiB {
+		klog.Infof("increasing requested disk size from %vGiB to %vGiB (minimal disk size)", requestGiB, consts.PerformancePlusMinimumDiskSizeGiB)
```
Comment (andyzhangx) [inline pkg/azuredisk/controllerserver_v2.go:98]: it should be `MinimumDiskSizeGiB`

**Review by andyzhangx — APPROVED** [review]
/lgtm
/retest

**Pull request merged**